<a href="https://colab.research.google.com/github/adnaksbhat/Colab_JupyterNotebooks/blob/main/wow.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Impacted Items Notifier using ECN

##Impacted notifier

In [ ]:
import pandas as pd

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

df=pd.read_excel('/content/All_ECN.xlsx')

In [ ]:
df_ecr=pd.read_excel('/content/All_ECR.xlsx')

In [ ]:
god_cols=['ECR_CODE','COSTING_PROJECTS','CHANGE DESCRIPTION','IMPACTEDITEM','DIVISION']

In [ ]:
df_ecr=df_ecr[god_cols]

In [ ]:
df['COSTING_PROJECTS']=df['COSTING_PROJECTS'].astype(str)
df['ECN_DESC']=df['ECN_DESC'].astype(str)
df['IMPACTED_ITEMS']=df['IMPACTED_ITEMS'].astype(str)

df_ecr['COSTING_PROJECTS']=df_ecr['COSTING_PROJECTS'].astype(str)
df_ecr['CHANGE DESCRIPTION']=df_ecr['CHANGE DESCRIPTION'].astype(str)
df_ecr['IMPACTEDITEM']=df_ecr['IMPACTEDITEM'].astype(str)

In [ ]:
df_ecr['CHANGE DESCRIPTION'] = df_ecr['CHANGE DESCRIPTION'].str.replace('_x000D_', '')

In [ ]:
df_ecr['CHANGE DESCRIPTION'] = df_ecr['CHANGE DESCRIPTION'].str.replace('\n', ' ')

In [ ]:
df['ECN_DESC']=df['ECN_DESC'].str.replace('\n',' ')
df['ECN_DESC']=df['ECN_DESC'].str.replace('_x000D_', '')

In [ ]:
df_ecr.head()

ECR_CODE COSTING_PROJECTS  \
0  ECR-356880          1209501   
1  ECR-356877          5847426   
2  ECR-356868          1606025   
3  ECR-356867          1NG0216   
4  ECR-356862           506053   

                                  CHANGE DESCRIPTION             IMPACTEDITEM  \
0                    revised after customer comments           SOA1513365/001   
1                  Remove Item Eng in BOM SMO0398124    M67508006_5847426/003   
2       Flushing kit piping arrangement modification       BOM_SMO0417502/000   
3  Add manual valves under section "piping and va...  BOM_LCOIPHRS0241102/000   
4                         Revisione MR bolts/gaskets           SKO2103175/001   

            DIVISION  
0  OTHER ASSIGNMENTS  
1                UPG  
2                 GT  
3  OTHER ASSIGNMENTS  
4        PIPE-LAYOUT

In [ ]:
df.head()

ECN_CODE                           ECN_TYPE  \
0  RCID113996  Item Revision Change Notice(RCID)   
1  CIS-019844                Scope Change Notice   
2  CIS-019845                Scope Change Notice   
3  RCID113995  Item Revision Change Notice(RCID)   
4  RCID113997  Item Revision Change Notice(RCID)   

                                            ECN_DESC COMMENT COSTING_PROJECTS  \
0  Sow is defined and hrs are allocated. so doc i...     NaN          1106653   
1                                 APPROVED---> 1 CIS     NaN          5841560   
2                  update spare parts ad per 1209552     NaN          1209567   
3  Sow is defined and hrs are enabled for this jo...     NaN          1106654   
4  VDR Document needs to be updated with customer...     NaN          1107551   

  PROBLEM_ITEMS                               IMPACTED_ITEMS  \
0           NaN                               SOS4578300/000   
1           NaN  S34105970_5841560/000,S34105970_5841560/001   
2           NaN  P009A0001_1209567/000,P000BM999_1209567/000   
3           NaN                               SOS4578301/000   
4           NaN                               SOS4578332/000   

                                SOLUTION_ITEMS ECR OFFICE Name  
0                               SOS4578300/001          SYS-OP  
1  S34105970_5841560/003,S34105970_5841560/002             UPG  
2  P000BM999_1209567/001,P009A0001_1209567/001             PVS  
3                               SOS4578301/001          SYS-OP  
4                               SOS4578332/001          SYS-OP

In [ ]:
import pandas as pd

def merge_ecr_to_df(df, df_ecr):
    """
    Merges rows of df_ecr to df based on matching criteria.
    """

    # Create temporary merge keys in both DataFrames
    df['merge_key'] = df['COSTING_PROJECTS'].astype(str) + df['ECN_DESC'].astype(str)
    df_ecr['merge_key'] = df_ecr['COSTING_PROJECTS'].astype(str) + df_ecr['CHANGE DESCRIPTION'].astype(str)

    # Merge the DataFrames on the merge key
    merged_df = pd.merge(df, df_ecr, on='merge_key', how='left', suffixes=('', '_ecr'))

    # Drop the temporary merge key column
    merged_df.drop(columns=['merge_key'], inplace=True)

    return merged_df

# Example usage
merged_df = merge_ecr_to_df(df, df_ecr)


In [ ]:
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 102646 entries, 0 to 102645
Data columns (total 14 columns):
 #   Column                Non-Null Count   Dtype 
---  ------                --------------   ----- 
 0   ECN_CODE              102646 non-null  object
 1   ECN_TYPE              102646 non-null  object
 2   ECN_DESC              102646 non-null  object
 3   COMMENT               46 non-null      object
 4   COSTING_PROJECTS      102646 non-null  object
 5   PROBLEM_ITEMS         117 non-null     object
 6   IMPACTED_ITEMS        102646 non-null  object
 7   SOLUTION_ITEMS        100603 non-null  object
 8   ECR OFFICE Name       68871 non-null   object
 9   ECR_CODE              30248 non-null   object
 10  COSTING_PROJECTS_ecr  30248 non-null   object
 11  CHANGE DESCRIPTION    30248 non-null   object
 12  IMPACTEDITEM          30248 non-null   object
 13  DIVISION              30248 non-null   object
dtypes: object(14)
memory usage: 11.0+ MB


In [ ]:
merged_df['ECN_CODE']=merged_df['ECN_CODE'].astype(str)
merged_df['ECN_TYPE']=merged_df['ECN_TYPE'].astype(str)
merged_df['ECN_DESC']=merged_df['ECN_DESC'].astype(str)
merged_df['COMMENT']=merged_df['COMMENT'].astype(str)
merged_df['COSTING_PROJECTS']=merged_df['COSTING_PROJECTS'].astype(str)
merged_df['PROBLEM_ITEMS']=merged_df['PROBLEM_ITEMS'].astype(str)
merged_df['IMPACTED_ITEMS']=merged_df['IMPACTED_ITEMS'].astype(str)
merged_df['SOLUTION_ITEMS']=merged_df['SOLUTION_ITEMS'].astype(str)

merged_df['ECR OFFICE Name']=merged_df['ECR OFFICE Name'].astype(str)
merged_df['ECR_CODE']=merged_df['ECR_CODE'].astype(str)
merged_df['COSTING_PROJECTS_ecr']=merged_df['COSTING_PROJECTS_ecr'].astype(str)
merged_df['CHANGE DESCRIPTION']=merged_df['CHANGE DESCRIPTION'].astype(str)
merged_df['IMPACTEDITEM']=merged_df['IMPACTEDITEM'].astype(str)
merged_df['DIVISION']=merged_df['DIVISION'].astype(str)



In [ ]:
merged_df["CHANGE DESCRIPTION"] = merged_df["CHANGE DESCRIPTION"].str.replace('\n', ' ')

In [ ]:
merged_df.head()

ECN_CODE                           ECN_TYPE  \
0  RCID113996  Item Revision Change Notice(RCID)   
1  CIS-019844                Scope Change Notice   
2  CIS-019845                Scope Change Notice   
3  RCID113995  Item Revision Change Notice(RCID)   
4  RCID113997  Item Revision Change Notice(RCID)   

                                            ECN_DESC COMMENT COSTING_PROJECTS  \
0  Sow is defined and hrs are allocated. so doc i...     nan          1106653   
1                                 APPROVED---> 1 CIS     nan          5841560   
2                  update spare parts ad per 1209552     nan          1209567   
3  Sow is defined and hrs are enabled for this jo...     nan          1106654   
4  VDR Document needs to be updated with customer...     nan          1107551   

  PROBLEM_ITEMS                               IMPACTED_ITEMS  \
0           nan                               SOS4578300/000   
1           nan  S34105970_5841560/000,S34105970_5841560/001   
2           nan  P009A0001_1209567/000,P000BM999_1209567/000   
3           nan                               SOS4578301/000   
4           nan                               SOS4578332/000   

                                SOLUTION_ITEMS ECR OFFICE Name    ECR_CODE  \
0                               SOS4578300/001          SYS-OP         nan   
1  S34105970_5841560/003,S34105970_5841560/002             UPG  ECR-356849   
2  P000BM999_1209567/001,P009A0001_1209567/001             PVS         nan   
3                               SOS4578301/001          SYS-OP         nan   
4                               SOS4578332/001          SYS-OP         nan   

  COSTING_PROJECTS_ecr  CHANGE DESCRIPTION           IMPACTEDITEM DIVISION  
0                  nan                 nan                    nan      nan  
1              5841560  APPROVED---> 1 CIS  S34105970_5841560/001      UPG  
2                  nan                 nan                    nan      nan  
3                  nan                 nan                    nan      nan  
4                  nan                 nan                    nan      nan

In [ ]:
def find_missing_impacts(row):

  impacted_items = str(row['IMPACTED_ITEMS']).split(',')  # Split comma-separated values
  impacted_item = str(row['IMPACTEDITEM'])
  missing = [item for item in impacted_items if item != impacted_item and item != 'nan']
  # Find items not in 'IMPACTEDITEM' and not 'nan'

  return ','.join(missing)  # Join missing impacts with commas





merged_df['missing_impacts'] = merged_df.apply(find_missing_impacts, axis=1)


In [ ]:
merged_df[merged_df['ECN_CODE']=='CIS-019854']

ECN_CODE             ECN_TYPE  \
7853  CIS-019854  Scope Change Notice   

                                               ECN_DESC COMMENT  \
7853  Aggiungere il dummy del bilancino M11101534 so...     nan   

     COSTING_PROJECTS PROBLEM_ITEMS         IMPACTED_ITEMS  \
7853          1704047           nan  M111AE999_1604379/002   

             SOLUTION_ITEMS ECR OFFICE Name    ECR_CODE COSTING_PROJECTS_ecr  \
7853  M111AE999_1604379/003             nan  ECR-356851              1704047   

                                     CHANGE DESCRIPTION  \
7853  Aggiungere il dummy del bilancino M11101534 so...   

               IMPACTEDITEM DIVISION        missing_impacts  
7853  M111AE999_1704047/012     CORE  M111AE999_1604379/002

In [ ]:
print(merged_df[merged_df['ECN_CODE']=='CIS-019854'].to_string())

        ECN_CODE             ECN_TYPE                                                                                            ECN_DESC COMMENT COSTING_PROJECTS PROBLEM_ITEMS         IMPACTED_ITEMS         SOLUTION_ITEMS ECR OFFICE Name    ECR_CODE COSTING_PROJECTS_ecr                                                                                  CHANGE DESCRIPTION           IMPACTEDITEM DIVISION        missing_impacts
7853  CIS-019854  Scope Change Notice  Aggiungere il dummy del bilancino M11101534 sotto il BOP Special Tools M111AE999 Ref. WF#202827089     nan          1704047           nan  M111AE999_1604379/002  M111AE999_1604379/003             nan  ECR-356851              1704047  Aggiungere il dummy del bilancino M11101534 sotto il BOP Special Tools M111AE999 Ref. WF#202827089  M111AE999_1704047/012     CORE  M111AE999_1604379/002


In [ ]:
print(df_ecr[df_ecr['ECR_CODE']=='ECR-356732'].to_string())

      ECR_CODE COSTING_PROJECTS                                         CHANGE DESCRIPTION   IMPACTEDITEM           DIVISION                                                         merge_key
29  ECR-356732          1NG0209  ajout: banc de maintenance bunle caisse à outils équippée  CLMST0244/000  OTHER ASSIGNMENTS  1NG0209ajout: banc de maintenance bunle caisse à outils équippée


In [ ]:
merged_df['COSTING_PROJECTS'] = merged_df['COSTING_PROJECTS'].astype(str)

In [ ]:
!pip install fuzzywuzzy[speedup]

##search method

In [ ]:
from __future__ import division
import ipywidgets as widgets
from IPython.display import display

import pandas as pd
import numpy as np
import re
import unicodedata
from fuzzywuzzy import fuzz

# Create input widgets
costing_projects_input = widgets.Text(description="JobNo:")
change_description_input = widgets.Text(description="Description:")
impacted_items_input = widgets.Text(description="Impacted Items:")
#division_input = widgets.Text(description="Division:")

search_button = widgets.Button(description="Search")

# Create output widget
output_widget = widgets.Output()



def normalize_text(text):
    text = unicodedata.normalize("NFKD", text)  # Normalize to remove special characters
    text = re.sub(r'\s+', ' ', text)  # Replace multiple spaces with a single space
    text = text.strip()  # Remove leading/trailing spaces
    return text

merged_df["CHANGE DESCRIPTION"] = merged_df["CHANGE DESCRIPTION"].astype(str).apply(normalize_text)
merged_df["IMPACTED_ITEMS"] = merged_df["IMPACTED_ITEMS"].astype(str).apply(normalize_text)
merged_df["COSTING_PROJECTS"] = merged_df["COSTING_PROJECTS"].astype(str).apply(normalize_text)


# Define search function
def search_df(button):
  with output_widget:
    output_widget.clear_output()
    costing_projects = str(costing_projects_input.value)
    change_description = normalize_text(change_description_input.value).replace('\n', '')
    #division=division_input.value
    impacted_items = str(impacted_items_input.value)






    # Filter the DataFrame
    pattern = '|'.join(impacted_items.split(','))  # Create regex pattern for any of the impacted items
    filtered_df = merged_df[
        (merged_df["COSTING_PROJECTS"] == costing_projects) &
        #(merged_df["CHANGE DESCRIPTION"] == change_description) &
        (merged_df["CHANGE DESCRIPTION"].apply(lambda x: fuzz.ratio(normalize_text(x), change_description) >= 90)) &
        #(merged_df["DIVISION"]==division)&
        (merged_df["IMPACTED_ITEMS"].str.contains(pattern, regex=True))
    ]



    if not filtered_df.empty:
          impacted_items_list = impacted_items.split(',')  # Split comma-separated values
          all_impacted_items = []
          for index in filtered_df.index:
            all_impacted_items.extend(str(filtered_df.loc[index, 'IMPACTED_ITEMS']).split(','))  # Get all impacted items from filtered rows

          # Find items not in impacted_items_list and not 'nan'
          remaining_missing = [item for item in all_impacted_items if item not in impacted_items_list and item != 'nan']
          remaining_missing = list(set(remaining_missing))  # Remove duplicates

          print(f"Remaining Missing Impacts: {','.join(remaining_missing)}")  # Join missing impacts with commas
    else:
          print("No matching records found.")
# search button

search_button.on_click(search_df)

# Display widgets


##output

In [ ]:
merged_df[(merged_df['COSTING_PROJECTS']=='1NG0218') & (merged_df['IMPACTEDITEM']=='ICSSH0031102/001')]

ECN_CODE                           ECN_TYPE  \
6649  RCID112651  Item Revision Change Notice(RCID)   

                                               ECN_DESC COMMENT  \
6649  Filter element material: stainless steel 316. ...     nan   

     COSTING_PROJECTS PROBLEM_ITEMS  \
6649          1NG0218           nan   

                                         IMPACTED_ITEMS  \
6649  ICSSH0031102/001,ICSSH1011102/001,ICSSH2011102...   

                                         SOLUTION_ITEMS ECR OFFICE Name  \
6649  ICSSH2011102/002,ICSSH1011102/002,ICSSH1021102...             ASE   

        ECR_CODE COSTING_PROJECTS_ecr  \
6649  ECR-352610              1NG0218   

                                     CHANGE DESCRIPTION      IMPACTEDITEM  \
6649  Filter element material: stainless steel 316. ...  ICSSH0031102/001   

               DIVISION                                    missing_impacts  
6649  OTHER ASSIGNMENTS  ICSSH1011102/001,ICSSH2011102/001,ICSSH1021102...

In [ ]:
display(costing_projects_input, change_description_input, impacted_items_input, search_button, output_widget)

Text(value='', description='JobNo:')

Text(value='', description='Description:')

Text(value='', description='Impacted Items:')

Button(description='Search', style=ButtonStyle())

Output()

In [ ]:
'''
1606624

1. Gas detector rack dimensions are updated in order to avoid interference\n2. Flowmeter VDR updated fro customer comments


1704047

Aggiungere il dummy del bilancino M11101534 sotto il BOP Special Tools M111AE999Ref. WF#202827089



5849119
due to Main motor shaft modification, we need to make the coupling guard shorter (20 mm); to avoid issue on casting body, we prefer to modify the axial dilatation joint

1604379
Change required to update P&ID with Customer comments
SOM6906969/001



1NG0218
Filter element material: stainless steel 316.
Minimum temperature changed, already discussed with supplier no material impact.
ICSSH0031102/001

'''

In [ ]:
from google.colab import files

df.to_excel('ECN_ECR_wow.xlsx', index=False)

files.download('ECN_ECR_wow.xlsx')

# Discipline | Activity

##data processing

In [2]:
import pandas as pd

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

df=pd.read_excel('/content/All_ECR.xlsx')

In [3]:
df.head(2)

ECR_CODE                    ECR_TYPE              SOURCE OF CHANGE  \
0  ECR-356880  Requisition Change Request  CHANGE DRIVEN BY ENGINEERING   
1  ECR-356877  Requisition Change Request  CHANGE DRIVEN BY ENGINEERING   

                  REASON_DESC  COMMENT  ACKNOWLEDGE_DATE  PROMISE_DATE  \
0  DESIGN PROGRESS COMPLETION      NaN               NaN           NaN   
1                TYPO MISTAKE      NaN               NaN           NaN   

      SSO ID ECR_CREATION_DATE    CHANGE_TYPE ENG IMPACTED LIST  \
0  105053425        2019-12-30     FFF&P ONLY               NaN   
1  212362330        2019-12-30  CIS/CIS+FFF&P               NaN   

  COSTING_PROJECTS               MATERIAL_CATEGORY   CHANGE_CATEGORY  \
0          1209501  EQUIPMENT/MATERIAL FULL SUPPLY          DOCUMENT   
1          5847426                           OTHER  REMOVE COMPONENT   

  CHANGE_ON_RECEIVED_ITEMS IMPACT VERIFICATION  \
0                       No                 Yes   
1                       No                 Yes   

                  CHANGE DESCRIPTION           IMPACTEDITEM  \
0    revised after customer comments         SOA1513365/001   
1  Remove Item Eng in BOM SMO0398124  M67508006_5847426/003   

         IMPACTED ITEM DESCRIPTION  FUNCTION           DIVISION  \
0  SKID INSTRUMENTATION LAYOUT DWG  IVPG-ENG  OTHER ASSIGNMENTS   
1      BOP - GT READY FOR SHIPMENT   GEP-ENG                UPG   

   ECR CREATION YEAR  AGING   APPROVER APPROVER_DECISION APPROVER DESC DATE  \
0               2019    1.5  105053425           APPROVE         2019-12-30   
1               2019    1.5  212362330           APPROVE         2019-12-30   

  CLOSURE DISPOSITION  MATURITY AS_BUILT_FROM_SITE REFERENCE ECR  \
0  Closed    Approved  Complete                 NO           NaN   
1  Closed    Approved  Complete                 NO           NaN   

  INTERNAL_REFERENT EXCLUDED EXCLUDED for Graph DEFECTIVE CHANGE       FW  \
0               NaN        N                  N              Yes  2019-53   
1               NaN        N                  N              Yes  2019-53   

                            ORGANIZATION NAME MAIN_FUNCTION   PL SUB_PL  \
0        IEP-IVPG-ENG-Pumps, Valves & Systems           ENG  TPS    PVS   
1  IEP-GEP-ENG-Project Engineering Office UPG           ENG  TPS    PEO   

          MANAGER NAME  ECR CREATION MONTH  
0  Guaglianone, Giulio                  12  
1      Osinski, Marcin                  12

In [4]:
good_cols=['ECR_CODE','CHANGE_TYPE','COSTING_PROJECTS','IMPACTEDITEM','IMPACTED ITEM DESCRIPTION','DIVISION','ENG IMPACTED LIST','CHANGE_CATEGORY','CHANGE DESCRIPTION']

In [5]:
df=df[good_cols]

In [6]:
df['IMPACTEDITEM_new']=df['IMPACTEDITEM']

<ipython-input-6-2ee7de3ae136>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['IMPACTEDITEM_new']=df['IMPACTEDITEM']


In [7]:
df["IMPACTEDITEM"] = df["IMPACTEDITEM"].astype(str)
df["IMPACTEDITEM_new"] = df["IMPACTEDITEM_new"].astype(str)

<ipython-input-7-39cf13568135>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["IMPACTEDITEM"] = df["IMPACTEDITEM"].astype(str)
<ipython-input-7-39cf13568135>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["IMPACTEDITEM_new"] = df["IMPACTEDITEM_new"].astype(str)


In [8]:
df['IMPACTEDITEM_new']=(df['IMPACTEDITEM_new'].str.split("/").str[0].str.replace('BOM_', ''))
df['IMPACTEDITEM_new']=(df['IMPACTEDITEM_new'].str.split("-").str[0])

In [9]:
df.head(2)

ECR_CODE    CHANGE_TYPE COSTING_PROJECTS           IMPACTEDITEM  \
0  ECR-356880     FFF&P ONLY          1209501         SOA1513365/001   
1  ECR-356877  CIS/CIS+FFF&P          5847426  M67508006_5847426/003   

         IMPACTED ITEM DESCRIPTION           DIVISION ENG IMPACTED LIST  \
0  SKID INSTRUMENTATION LAYOUT DWG  OTHER ASSIGNMENTS               NaN   
1      BOP - GT READY FOR SHIPMENT                UPG               NaN   

    CHANGE_CATEGORY                 CHANGE DESCRIPTION   IMPACTEDITEM_new  
0          DOCUMENT    revised after customer comments         SOA1513365  
1  REMOVE COMPONENT  Remove Item Eng in BOM SMO0398124  M67508006_5847426

## VDR

In [10]:
import pandas as pd

df_all_vdr=pd.read_excel("/content/All_VDR.xlsx")

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

df_all_vdr.head(2)

PROJECT_ID RESOURCE_NAME      ACTIVITY_ID  \
0     504040           NaN  061EMF01GA60701   
1    2846557     undefined          LL00M69   

                                       ACTIVITY_NAME       VDR_EARLY_DATE  \
0  Generator circuit breaker outline & foundation...  2001-05-08 00:00:00   
1   Schema Interconnessioni con Pannelo di Controllo  2007-09-10 00:00:00   

  PLANNING_DATE PROMISE_DATE COMMENT  ACKNOWLEDGE LAST MODIFIED DATE  \
0    2001-05-08          NaT     NaN          NaN                NaT   
1    2007-09-10          NaT     NaN          NaN                NaT   

            STATUS   REAL_CODE  \
0    COMPLETEDLATE  SOK4402871   
1  COMPLETEDONTIME    SOM43475   

                                        PROJECT_NAME   ROLE_ID  TYPE  \
0                     0504040-FGP FR9 Modularization  ELEPLANT   NaN   
1  2846557-SNAM RETE GAS SPA (adeguamento Atex Ca...      CTRL   NaN   

   ACT_TYPE ACTUAL_FINISH RESOURCE_SSO  ISSUE_NUMBER LAST_SUBMISSION  \
0     560.0    2014-05-08          NaN             0               N   
1     560.0    2007-09-10          NaN             0               N   

  FIRST_SUBMISSION ORGANIZATION_NAME P6-PROJ_CONTR_DELIVERY_DATE  \
0                Y               NaN                         NaN   
1                Y               NaN                         NaN   

  DOC_EMISSION_CODE   PL  SUM CONC  Open  Backlog  EXCLUDE LONG_DUMMY  \
0               NaN  NaN  NaN  NaN   NaN      NaN      NaN        NaN   
1               NaN  NaN  NaN  NaN   NaN      NaN      NaN        NaN   

  TRAIN_TYPE VDR_EARLY_DUE_WEEK   RETURN_DATE_ACTUAL RETURN_DATE_PLANNED  \
0        IPL                NaN  2014-06-11 00:00:00                 NaT   
1      GT+CC                NaN  2007-12-10 00:00:00          2007-09-24   

  UNIFIER_PROJECT_STATUS  DELAY  Completed Late AGING DOC_RETURN_CODE  \
0                 Active    NaN             NaN   NaN             NaN   
1               Inactive    NaN             NaN   NaN             NaN   

  DIVISION PLANNING_FW SYSTEM_DATE TEAM LEADER PLANNING QTR PLANNING YEAR  \
0      NaN         NaN         NaT         NaN          NaN           NaN   
1      NaN         NaN         NaT         NaN          NaN           NaN   

  Completed_On_Time SUB_PL               RAC_DT  DELAY_CUSTOMER  \
0               NaN    NaN  2001-05-08 00:00:00             NaN   
1               NaN    NaN  2007-09-10 00:00:00             NaN   

  PLANNING MONTH UNIFIER_REVISION_STATUS                       CUSTOMER_DOC  \
0            NaN                     NaN  052-9400-PPP-LAY-NPI-000-00013-08   
1            NaN                     NaN                                NaN   

  DOC_TYPE  PO_NUMBER UNIFIER_DOCUMENT_STATUS CONTRACT_NUMBER FUNCTION  \
0      NaN        NaN                     NaN       M_1708026      NaN   
1      NaN        NaN                     NaN       M_2846498      NaN   

  RETURN_CODE PROJECT_PHASE PO_SPECIFICATION_ID VENDOR SIGN-OFF_STATUS  \
0         NaN           NaN                 NaN    NaN             NaN   
1         NaN           NaN                 NaN    NaN             NaN   

  APPROVE_TYPE DOCUMENT_DUE_DATE VDR_DESIGN_PHASE ADDITIONAL_CUSTOMER_DOC  \
0          NaN               NaT               L2                     NaN   
1          NaN               NaT               L2                     NaN   

  ITEM_AGGREGATION JOB_TYPE MAIN_FUNCTION                           PE  \
0              NaN      NaN           NaN  Marronaro, Ms. Maria Grazia   
1              NaN      NaN           NaN        Sabatano, Mr. Roberto   

   Design_Configurator PROJECT_STATUS REVISION_NUMBER                     PM  \
0                  NaN            NaN             NaN    Lanciani, Mr. Carlo   
1                  NaN            NaN             NaN  La Valle, Mr. Antonio   

                  PP  STEP_ID  STEP_NAME FLAG GENERAL SUB_DIVISION  
0  Massimo Bortolani      NaN        NaN  NaN     NaN          NaN  
1    Mirko Miniatesi      NaN        NaN  NaN   

In [11]:
df_vdr=df_all_vdr[["PROJECT_ID","ACTIVITY_ID","ACTIVITY_NAME","REAL_CODE"]]

In [12]:
df_vdr.shape

(160264, 4)

In [13]:
df_vdr['PROJECT_ID']=df_vdr['PROJECT_ID'].astype(str)
df_vdr['ACTIVITY_ID']=df_vdr['ACTIVITY_ID'].astype(str)
df_vdr['ACTIVITY_NAME']=df_vdr['ACTIVITY_NAME'].astype(str)
df_vdr['REAL_CODE']=df_vdr['REAL_CODE'].astype(str)

<ipython-input-13-de3d78f8cf9f>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_vdr['PROJECT_ID']=df_vdr['PROJECT_ID'].astype(str)
<ipython-input-13-de3d78f8cf9f>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_vdr['ACTIVITY_ID']=df_vdr['ACTIVITY_ID'].astype(str)
<ipython-input-13-de3d78f8cf9f>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pyd

In [14]:
# prompt: if same 'PROJECT_ID' and 'ACT_ID' are repeated in another row then drop the rows with REAL_CODE as NaN

import pandas as pd

# Assuming your DataFrame is named 'df'
def drop_nan_duplicates(df):
    """Drops rows with NaN REAL_CODE where PROJECT_ID and ACT_ID are duplicated."""

    # Group by 'PROJECT_ID' and 'ACT_ID'
    grouped = df.groupby(['PROJECT_ID', 'ACTIVITY_ID'])

    # Iterate through each group
    rows_to_drop = []
    for (project_id, activity_id), group in grouped:
        if len(group) > 1:  # Check for duplicates
          # Identify rows with NaN REAL_CODE within the duplicate group
            nan_real_code_rows = group[group['REAL_CODE'].isna()]
            rows_to_drop.extend(nan_real_code_rows.index.tolist())

    # Drop identified rows from original dataframe
    df_cleaned = df.drop(index=rows_to_drop)
    return df_cleaned


df_vdr = drop_nan_duplicates(df_vdr)


In [15]:
df_vdr.shape

(160264, 4)

In [16]:
df_vdr.head()

PROJECT_ID      ACTIVITY_ID  \
0     504040  061EMF01GA60701   
1    2846557          LL00M69   
2    2846557          LL00M82   
3    2846557          LL00M86   
4    2846557          LL00M77   

                                       ACTIVITY_NAME   REAL_CODE  
0  Generator circuit breaker outline & foundation...  SOK4402871  
1   Schema Interconnessioni con Pannelo di Controllo    SOM43475  
2          Connessioni Interne Pannello di Controllo    SOM43541  
3                   Layout Quadro e lista componenti    SOM40611  
4                           Elenco Utenze Elettriche  SOK6774781

##ISPO data

In [17]:
import pandas as pd

df_all_ispo=pd.read_excel("/content/All_ISPO.xlsx")

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

df_all_ispo.head(2)

PROJECT_ID PRIMARY_RESOURCE_NAME   ACT_ID                  ACTIVITY_NAME  \
0    1102467                   NaN  PC00S30  CECO DESIGN REVIEW 1 - REPORT   
1    1102467                   NaN  PC00S30  CECO DESIGN REVIEW 1 - REPORT   

  STEP_ID STEP_NAME IMPACTED_NCR EARLY_DATE LATE_FINISH PROMISE_DATE COMMENT  \
0     NaN       NaN          NaN 2009-05-04         NaT          NaT     NaN   
1     NaN       NaN          NaN 2009-05-04         NaT          NaT     NaN   

   ACKNOWLEDGE LAST MODIFIED DATE               PROJECT_NAME    ROLE  \
0          NaN                NaT  1102467-ADCO SAS GAS LIFT  CCCALC   
1          NaN                NaT  1102467-ADCO SAS GAS LIFT  CCCALC   

  PRIMARY_RESOURCE_SSO LONG_DUMMY  ACT_TYPE   REAL_CODE   EARLY_STATUS  DELAY  \
0                  NaN        NaN     510.0  SOS8629724  COMPLETEDLATE    NaN   
1                  NaN        NaN     510.0         NaN  COMPLETEDLATE    NaN   

  ACTUAL_FINISH      TYPE EARLY_DUE_WEEK AGING ITEM_AGGREGATION  \
0    2009-05-13  Activity             18   NaN              NaN   
1    2009-05-13      Step             18   NaN              NaN   

  P6-PROJ_CONTR_DELIVERY_DATE   PL  PO_NUMBER  PROJECT_MANAGER  \
0                         NaN  NaN        NaN              NaN   
1                         NaN  NaN        NaN              NaN   

   Completed Late               RAC_DT PLANNING_DATE TRAIN_TYPE SUB_PL  \
0             NaN  2009-02-17 00:00:00           NaT        NaN    NaN   
1             NaN  2009-02-17 00:00:00           NaT        NaN    NaN   

  ORGANIZATION_NAME SYSTEM_DATE PROJECT_PHASE CONTRACT_NUMBER JOB_TYPE  \
0               NaN         NaT        CLOSED       M_1102467      NaN   
1               NaN         NaT        CLOSED       M_1102467      NaN   

  DIVISION CONC  Backlog  EXCLUDE FUNCTION  PROJECT_ENGINEER  PROJECT_PLANNER  \
0      NaN  NaN      NaN      NaN      NaN               NaN              NaN   
1      NaN  NaN      NaN      NaN      NaN               NaN              NaN   

  DOC_TYPE LATE_STATUS MAIN_FUNCTION  Open PLANNING_FW UNIFIER_PROJECT_STATUS  \
0      NaN         NaN           NaN   NaN         NaN                    NaN   
1      NaN         NaN           NaN   NaN         NaN                    NaN   

   Completed_On_Time  SUM PO_SPECIFICATION_ID VENDOR SIGN-OFF_STATUS  \
0                NaN  NaN                 NaN    NaN             NaN   
1                NaN  NaN                 NaN    NaN             NaN   

  APPROVE_TYPE DOCUMENT_DUE_DATE  DWG_ISSUE_NUMBER_FINAL Design_Configurator  \
0          NaN               NaT                     NaN                 NaN   
1          NaN               NaT                     NaN                 NaN   

            PM  RETURN_DATE_PLANNED  BL_Project_Finish  RETURN_CODE  \
0  Tariq Ahmed                  NaN                NaN          NaN   
1  Tariq Ahmed                  NaN                NaN          NaN   

   DOC_RETURN_CODE  ACT_REGIONAL_ERP  ADDITIONAL_CUSTOMER_DOC  \
0              NaN               NaN                      NaN   
1              NaN               NaN                      NaN   

                      PE  ACT_AGGREGATION  GM  RETURN_DATE_ACTUAL  \
0  Laura Fernanda Grayeb              NaN NaN                 NaN   
1  Laura Fernanda Grayeb              NaN NaN                 NaN   

  PROJECT_STATUS_Primavera  FIRST_SUBMISSION  LAST_SUBMISSION  DELAY_CUSTOMER  \
0                      NaN               NaN              NaN             NaN   
1                      NaN               NaN              NaN             NaN   

   CUSTOMER_DOC  UNIFIER_DOCUMENT_STATUS  UNIFIER_REVISION_STATUS  \
0           NaN                      NaN                      NaN   
1           NaN                      NaN                      NaN   

   PLANNING_YEAR PLANNING_MONTH PLANNING QTR TEAM LEADER  ACT_DEPARTMENT  \
0            NaN            NaN          NaN         NaN             NaN   
1            NaN            NaN          NaN      

In [18]:
df_ispo=df_all_ispo[["PROJECT_ID","ACT_ID","ACTIVITY_NAME","LONG_DUMMY","REAL_CODE"]]

In [19]:
df_ispo.shape

(709450, 5)

In [20]:
# prompt: if same 'PROJECT_ID' and 'ACT_ID' are repeated in another row then drop the rows with REAL_CODE as NaN

import pandas as pd

# Assuming your DataFrame is named 'df'
def drop_nan_duplicates(df):
    """Drops rows with NaN REAL_CODE where PROJECT_ID and ACT_ID are duplicated."""

    # Group by 'PROJECT_ID' and 'ACT_ID'
    grouped = df.groupby(['PROJECT_ID', 'ACT_ID'])

    # Iterate through each group
    rows_to_drop = []
    for (project_id, act_id), group in grouped:
        if len(group) > 1:  # Check for duplicates
          # Identify rows with NaN REAL_CODE within the duplicate group
            nan_real_code_rows = group[group['REAL_CODE'].isna()]
            rows_to_drop.extend(nan_real_code_rows.index.tolist())

    # Drop identified rows from original dataframe
    df_cleaned = df.drop(index=rows_to_drop)
    return df_cleaned


df_ispo = drop_nan_duplicates(df_ispo)


In [21]:
df_ispo.shape

(627018, 5)

In [22]:
df_ispo.head()

PROJECT_ID   ACT_ID                  ACTIVITY_NAME LONG_DUMMY   REAL_CODE
0     1102467  PC00S30  CECO DESIGN REVIEW 1 - REPORT        NaN  SOS8629724
2     1102467  PC00U50            BASIC DESIGN REVIEW        NaN  SOK6747811
4     1102467  SP00S21      ROTATION SENSE DEFINITION        NaN  SOS9989279
6     1102467  DS00S42           FUNCTIONAL DATA FORM        NaN  SOK7352503
10    1102467  DS00S55         CC INTERNAL DATA SHEET        NaN  SOS9989297

In [23]:
# prompt: in df2 if 'REAL_CODE' is null or has 'DELETED IN TCE' then replace it with its corresponding 'LONG_DUMMY' value. if LONG_DUMMY value is also null then fill both 'REAL_CODE' and "LONG_DUMMY' as 'unknown'

# Assuming your DataFrame is named 'df2'
def fill_real_code(df2):
    for index, row in df2.iterrows():
        if pd.isnull(row['REAL_CODE']):
            if pd.notnull(row['LONG_DUMMY']) and row['LONG_DUMMY'] != '':
                df2.loc[index, 'REAL_CODE'] = row['LONG_DUMMY']
            else:
                df2.loc[index, 'REAL_CODE'] = 'unknown'
                df2.loc[index, 'LONG_DUMMY'] = 'unknown'
    return df2

df_ispo_2 = fill_real_code(df_ispo)

## merge ECR and VDR and ISPO

In [24]:
df_ecr_vdr = pd.merge(df, df_vdr,  left_on=['IMPACTEDITEM_new'], right_on=['REAL_CODE'], how='left')




In [26]:
df_ecr_vdr.head()

ECR_CODE    CHANGE_TYPE COSTING_PROJECTS             IMPACTEDITEM  \
0  ECR-356880     FFF&P ONLY          1209501           SOA1513365/001   
1  ECR-356877  CIS/CIS+FFF&P          5847426    M67508006_5847426/003   
2  ECR-356868     FFF&P ONLY          1606025       BOM_SMO0417502/000   
3  ECR-356867     FFF&P ONLY          1NG0216  BOM_LCOIPHRS0241102/000   
4  ECR-356862     FFF&P ONLY           506053           SKO2103175/001   

                   IMPACTED ITEM DESCRIPTION           DIVISION  \
0            SKID INSTRUMENTATION LAYOUT DWG  OTHER ASSIGNMENTS   
1                BOP - GT READY FOR SHIPMENT                UPG   
2         MS5001 TURBOGENERATOR FLUSHING KIT                 GT   
3        PIPING AND VALVES FOR RUN DOWN TANK  OTHER ASSIGNMENTS   
4  MATERIAL REQUISITION G0201 BOLTS AND NUTS        PIPE-LAYOUT   

  ENG IMPACTED LIST   CHANGE_CATEGORY  \
0               NaN          DOCUMENT   
1               NaN  REMOVE COMPONENT   
2               NaN        CHANGE QTY   
3               NaN     ADD COMPONENT   
4               NaN        CHANGE QTY   

                                  CHANGE DESCRIPTION   IMPACTEDITEM_new  \
0                    revised after customer comments         SOA1513365   
1                  Remove Item Eng in BOM SMO0398124  M67508006_5847426   
2       Flushing kit piping arrangement modification         SMO0417502   
3  Add manual valves under section "piping and va...    LCOIPHRS0241102   
4                         Revisione MR bolts/gaskets         SKO2103175   

  PROJECT_ID ACTIVITY_ID ACTIVITY_NAME REAL_CODE  
0        NaN         NaN           NaN       NaN  
1        NaN         NaN           NaN       NaN  
2        NaN         NaN           NaN       NaN  
3        NaN         NaN           NaN       NaN  
4        NaN         NaN           NaN       NaN

In [27]:
df_ecr_vdr = df_ecr_vdr.rename(columns={'ACTIVITY_ID': 'ACTIVITY_ID_VDR','ACTIVITY_NAME':'ACTIVITY_NAME_VDR','PROJECT_ID':'PROJECT_ID_VDR'})

In [28]:
df_ecr_vdr.shape

(40908, 14)

In [25]:
from google.colab import files
df_ecr_vdr.to_excel('df_ecr_vdr.xlsx', index=False)
files.download('df_ecr_vdr.xlsx')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [29]:
df_ecr_vdr_ispo = pd.merge(df_ecr_vdr, df_ispo,  left_on=['IMPACTEDITEM_new'], right_on=['REAL_CODE'], how='left')

In [30]:
df_ecr_vdr_ispo.head()

ECR_CODE    CHANGE_TYPE COSTING_PROJECTS             IMPACTEDITEM  \
0  ECR-356880     FFF&P ONLY          1209501           SOA1513365/001   
1  ECR-356877  CIS/CIS+FFF&P          5847426    M67508006_5847426/003   
2  ECR-356868     FFF&P ONLY          1606025       BOM_SMO0417502/000   
3  ECR-356867     FFF&P ONLY          1NG0216  BOM_LCOIPHRS0241102/000   
4  ECR-356862     FFF&P ONLY           506053           SKO2103175/001   

                   IMPACTED ITEM DESCRIPTION           DIVISION  \
0            SKID INSTRUMENTATION LAYOUT DWG  OTHER ASSIGNMENTS   
1                BOP - GT READY FOR SHIPMENT                UPG   
2         MS5001 TURBOGENERATOR FLUSHING KIT                 GT   
3        PIPING AND VALVES FOR RUN DOWN TANK  OTHER ASSIGNMENTS   
4  MATERIAL REQUISITION G0201 BOLTS AND NUTS        PIPE-LAYOUT   

  ENG IMPACTED LIST   CHANGE_CATEGORY  \
0               NaN          DOCUMENT   
1               NaN  REMOVE COMPONENT   
2               NaN        CHANGE QTY   
3               NaN     ADD COMPONENT   
4               NaN        CHANGE QTY   

                                  CHANGE DESCRIPTION   IMPACTEDITEM_new  \
0                    revised after customer comments         SOA1513365   
1                  Remove Item Eng in BOM SMO0398124  M67508006_5847426   
2       Flushing kit piping arrangement modification         SMO0417502   
3  Add manual valves under section "piping and va...    LCOIPHRS0241102   
4                         Revisione MR bolts/gaskets         SKO2103175   

  PROJECT_ID_VDR ACTIVITY_ID_VDR ACTIVITY_NAME_VDR REAL_CODE_x PROJECT_ID  \
0            NaN             NaN               NaN         NaN        NaN   
1            NaN             NaN               NaN         NaN        NaN   
2            NaN             NaN               NaN         NaN    1606025   
3            NaN             NaN               NaN         NaN    1NG0214   
4            NaN             NaN               NaN         NaN     506053   

      ACT_ID                                  ACTIVITY_NAME  \
0        NaN                                            NaN   
1        NaN                                            NaN   
2  M11101941          ENG-FLUSHING KIT FOR MINERAL LUBE OIL   
3  S61000640        ENG PIPING AND VALVES FOR RUN DOWN TANK   
4  K005PTA31  ENG - BOLTS AND BOLTS AND NUTS. - BALANCE MTO   

             LONG_DUMMY      REAL_CODE_y  
0                   NaN              NaN  
1                   NaN              NaN  
2  M11101941010_1606025       SMO0417502  
3  S61000640010_1NG0214  LCOIPHRS0241102  
4  K005PTA31010_0506053       SKO2103175

In [31]:
df_ecr_vdr_ispo = df_ecr_vdr_ispo.rename(columns={'ACT_ID': 'ACTIVITY_ID_ISPO','ACTIVITY_NAME':'ACTIVITY_NAME_ISPO','PROJECT_ID':'PROJECT_ID_ISPO','REAL_CODE_x':'REAL_CODE_VDR','REAL_CODE_y':'REAL_CODE_ISPO'})

In [32]:
from google.colab import files
df_ecr_vdr_ispo.to_excel('df_ecr_vdr_ispo.xlsx', index=False)
files.download('df_ecr_vdr_ispo.xlsx')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

## group DIVISION wise and IMAPCTEDFILE wise and DISCIPLINE wise

In [33]:
import pandas as pd

df_final=pd.read_excel("/content/df_ecr_vdr_ispo.xlsx")

In [58]:
df_final.head(2)

ECR_CODE    JobNo           IMPACTEDITEM  \
0  ECR-356880  1209501         SOA1513365/001   
1  ECR-356877  5847426  M67508006_5847426/003   

         IMPACTED ITEM DESCRIPTION Impacted_Discipline ACTIVITY_ID_VDR  \
0  SKID INSTRUMENTATION LAYOUT DWG                 NaN             NaN   
1      BOP - GT READY FOR SHIPMENT                 NaN             NaN   

  ACTIVITY_NAME_VDR ACTIVITY_ID_ISPO ACTIVITY_NAME_ISPO Activity_Id  \
0               NaN              NaN                NaN         NaN   
1               NaN              NaN                NaN         NaN   

  Activity_Name  
0           NaN  
1           NaN

In [41]:
df_final = df_final.rename(columns={'COSTING_PROJECTS': 'JobNo','ENG IMPACTED LIST':'Impacted_Discipline'})

In [42]:
good_cols=['ECR_CODE','JobNo','IMPACTEDITEM','IMPACTED ITEM DESCRIPTION','Impacted_Discipline','ACTIVITY_ID_VDR','ACTIVITY_NAME_VDR','ACTIVITY_ID_ISPO','ACTIVITY_NAME_ISPO']

In [44]:
df_final=df_final[good_cols]

In [59]:
df_final.head(2)

ECR_CODE    JobNo           IMPACTEDITEM  \
0  ECR-356880  1209501         SOA1513365/001   
1  ECR-356877  5847426  M67508006_5847426/003   

         IMPACTED ITEM DESCRIPTION Impacted_Discipline ACTIVITY_ID_VDR  \
0  SKID INSTRUMENTATION LAYOUT DWG                 NaN             NaN   
1      BOP - GT READY FOR SHIPMENT                 NaN             NaN   

  ACTIVITY_NAME_VDR ACTIVITY_ID_ISPO ACTIVITY_NAME_ISPO Activity_Id  \
0               NaN              NaN                NaN         NaN   
1               NaN              NaN                NaN         NaN   

  Activity_Name  
0           NaN  
1           NaN

In [46]:
# prompt: in df_final create a new col 'Activity_Id'
# for each row it should take value from corresponding value of ACTIVITY_ID_VDR col. if null take from ACTIVITY_ID_ISPO col

df_final['Activity_Id'] = df_final['ACTIVITY_ID_VDR'].fillna(df_final['ACTIVITY_ID_ISPO'])
df_final['Activity_Name'] = df_final['ACTIVITY_NAME_VDR'].fillna(df_final['ACTIVITY_NAME_ISPO'])

In [62]:
df_final.shape

(47778, 11)

In [63]:
df_final.drop_duplicates(inplace=True)

In [64]:
df_final.shape

(40193, 11)

In [72]:
df_final['Activity_Id'].isnull().sum()

23763

In [73]:
df_final['ACT_IMPACTED']=df_final.groupby('Activity_Id')['IMPACTEDITEM'].transform(lambda x: ','.join(x.astype(str).unique()))

In [74]:
df_final['ACT_DISCIPLINE']=df_final.groupby('Activity_Id')['Impacted_Discipline'].transform(lambda x: ','.join(x.astype(str).unique()))

In [76]:
import pandas as pd

def get_unique_disciplines(df):
    """
    Extracts unique discipline values from the 'ACT_DISCIPLINE' column,
    handling potential comma-separated values within cells.
    """
    all_disciplines = []
    for value in df['ACT_DISCIPLINE']:
        if pd.isna(value): # handle NaN
            continue
        disciplines = str(value).split(',')
        for disc in disciplines:
            cleaned_disc = disc.strip()
            if cleaned_disc: #avoid empty string
                all_disciplines.append(cleaned_disc)

    unique_disciplines = list(set(all_disciplines))
    return unique_disciplines


unique_disciplines = get_unique_disciplines(df_final)




In [77]:
df_final['ACT_DISCIPLINE'] = df_final['ACT_DISCIPLINE'].apply(lambda x: ','.join(get_unique_disciplines(pd.DataFrame({'ACT_DISCIPLINE': [x]}))))

In [79]:
# prompt: if nan is present among ACT_DISCIPLINE value like 'VESS,ELE,CTRL-for aux,INSTR-for aux,RC F2F,nan,MECH' replace it with ''


def replace_nan_with_empty(value):
    if isinstance(value, str):
      return value.replace('nan', '')
    return value

df_final['ACT_DISCIPLINE'] = df_final['ACT_DISCIPLINE'].apply(replace_nan_with_empty)

In [82]:
df_final.head()

ECR_CODE    JobNo             IMPACTEDITEM  \
0  ECR-356880  1209501           SOA1513365/001   
1  ECR-356877  5847426    M67508006_5847426/003   
2  ECR-356868  1606025       BOM_SMO0417502/000   
3  ECR-356867  1NG0216  BOM_LCOIPHRS0241102/000   
4  ECR-356862   506053           SKO2103175/001   

                   IMPACTED ITEM DESCRIPTION Impacted_Discipline  \
0            SKID INSTRUMENTATION LAYOUT DWG                 NaN   
1                BOP - GT READY FOR SHIPMENT                 NaN   
2         MS5001 TURBOGENERATOR FLUSHING KIT                 NaN   
3        PIPING AND VALVES FOR RUN DOWN TANK                 NaN   
4  MATERIAL REQUISITION G0201 BOLTS AND NUTS                 NaN   

  ACTIVITY_ID_VDR ACTIVITY_NAME_VDR ACTIVITY_ID_ISPO  \
0             NaN               NaN              NaN   
1             NaN               NaN              NaN   
2             NaN               NaN        M11101941   
3             NaN               NaN        S61000640   
4             NaN               NaN        K005PTA31   

                              ACTIVITY_NAME_ISPO Activity_Id  \
0                                            NaN         NaN   
1                                            NaN         NaN   
2          ENG-FLUSHING KIT FOR MINERAL LUBE OIL   M11101941   
3        ENG PIPING AND VALVES FOR RUN DOWN TANK   S61000640   
4  ENG - BOLTS AND BOLTS AND NUTS. - BALANCE MTO   K005PTA31   

                                   Activity_Name  \
0                                            NaN   
1                                            NaN   
2          ENG-FLUSHING KIT FOR MINERAL LUBE OIL   
3        ENG PIPING AND VALVES FOR RUN DOWN TANK   
4  ENG - BOLTS AND BOLTS AND NUTS. - BALANCE MTO   

                                        ACT_IMPACTED ACT_DISCIPLINE  
0                                                NaN                 
1                                                NaN                 
2  BOM_SMO0417502/000,BOM_SSO2493917/003,BOM_SSO2...                 
3  BOM_LCOIPHRS0241102/000,SKO0061411/000,BOM_SKO...                 
4       SKO2103175/001,SKO2103174/000,SKO2103175/003

In [83]:
from google.colab import files
df_final.to_excel('Disciplines_impacted_all.xlsx', index=False)
files.download('Disciplines_impacted_all.xlsx')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

##Dont run this cell -> | multiple activity found

In [66]:
# get all rows where ACTIVITY_ID_ISPO and ACTIVITY_ID_VDR is not null and if they are same ignore it
filtered_df = df_final[df_final['ACTIVITY_ID_ISPO'].notnull() & df_final['ACTIVITY_ID_VDR'].notnull()]

#Further filter to remove rows where ACTIVITY_ID_ISPO and ACTIVITY_ID_VDR are the same
filtered_df = filtered_df[filtered_df['ACTIVITY_ID_ISPO'] != filtered_df['ACTIVITY_ID_VDR']]


In [67]:
filtered_df.shape

(934, 11)

In [69]:
filtered_df=filtered_df[['IMPACTEDITEM','ACTIVITY_ID_VDR','ACTIVITY_NAME_VDR','ACTIVITY_ID_ISPO','ACTIVITY_NAME_ISPO']]

In [70]:
filtered_df.head()

IMPACTEDITEM ACTIVITY_ID_VDR               ACTIVITY_NAME_VDR  \
107  SKO0060540/002         EX01U02  BASEPLATE CONSTRUCTION DRAWING   
202  SOM6808769/003         FD00M03  G.T. AUXILIARY FOUNDATION DWG.   
209  SOM6913204/000         OE00U00        LOCAL GAUGE BOARD LAYOUT   
215  SOK2313444/000         EX00U19  PAINTING SPECIFICATION FOR BOP   
217  SOK6631183/002         CD01U53               ISOMETRIC DRAWING   

     ACTIVITY_ID_ISPO                               ACTIVITY_NAME_ISPO  
107  S000BASEPLATEOIL                                    ENG-BASEPLATE  
202           FD00U60  FILTER HOUSE STRUCTURE FOUNDATION LOADS DRAWING  
209           OM00U10                  LUBE OIL CONSOLE INSTR. LAY-OUT  
215   002ABA01SP00K01           ENG DOC-GENERAL PAINTING SPECIFICATION  
217   028PTA01CD30K01                ENG DOC-PIPING ISOMETRIC DRAWINGS

In [71]:
from google.colab import files
filtered_df.to_excel('multiple_activity.xlsx', index=False)
files.download('multiple_activity.xlsx')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

## take only good cols and not null values

Total ECR rows -> **47089**


No. of rows with null for :

ActIds is :  **23763**

Eng Impacted List is : **35311**


==================
After grouping Disciplines and Impacted Items :
__________________
No . unique Activity ID  rows : **3460**

No. of rows with Disciplines as null : **2615**

So, total non null rows -> 845

__________________________




In [85]:
df_final['Activity_Id'].isnull().sum()

23763

In [87]:
df_final['Impacted_Discipline'].isnull().sum()

35311

In [89]:
df_final.head(2)

ECR_CODE    JobNo           IMPACTEDITEM  \
0  ECR-356880  1209501         SOA1513365/001   
1  ECR-356877  5847426  M67508006_5847426/003   

         IMPACTED ITEM DESCRIPTION Impacted_Discipline ACTIVITY_ID_VDR  \
0  SKID INSTRUMENTATION LAYOUT DWG                 NaN             NaN   
1      BOP - GT READY FOR SHIPMENT                 NaN             NaN   

  ACTIVITY_NAME_VDR ACTIVITY_ID_ISPO ACTIVITY_NAME_ISPO Activity_Id  \
0               NaN              NaN                NaN         NaN   
1               NaN              NaN                NaN         NaN   

  Activity_Name ACT_IMPACTED ACT_DISCIPLINE  
0           NaN          NaN                 
1           NaN          NaN

In [90]:
new_cols=['Activity_Id','Activity_Name','ACT_IMPACTED','ACT_DISCIPLINE']

In [91]:
dfdf=df_final[new_cols]

In [97]:
dfdf.shape

(5006, 4)

In [93]:
dfdf.drop_duplicates(inplace=True)

<ipython-input-93-9803d1c14469>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dfdf.drop_duplicates(inplace=True)


In [96]:
dfdf=dfdf[dfdf['Activity_Id'].isna()==False]

In [98]:
dfdf.tail(2)

Activity_Id                       Activity_Name  \
47741   M64008610  ELECTRICAL PACKAGING SPECIFICATION   
47770   N34A06584            ENG-MISCELLANEA FOR RING   

                                ACT_IMPACTED ACT_DISCIPLINE  
47741  BOM_SMO0126889/000,BOM_SMO0126629/000                 
47770                     BOM_SNO3704951/000

After grouping :

In [100]:
from google.colab import files
dfdf.to_excel('Disciplines_impacted_notnull.xlsx', index=False)
files.download('Disciplines_impacted_notnull.xlsx')


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

# Missing Discipline notifier

##data

In [1]:
import pandas as pd

df=pd.read_excel("/content/Disciplines_impacted_notnull.xlsx")

FileNotFoundError: [Errno 2] No such file or directory: '/content/Disciplines_impacted_notnull.xlsx'

In [2]:
df.head()

NameError: name 'df' is not defined

##Output

In [112]:
from __future__ import division
import ipywidgets as widgets
from IPython.display import display

import pandas as pd
import numpy as np
import re
import unicodedata
#from fuzzywuzzy import fuzz

# Create input widgets
activity_input = widgets.Text(description="Activity Id:")
impacted_discipline_input = widgets.Text(description="Disciplines impacted:")
impacted_items_input = widgets.Text(description="Impacted Items:")
#division_input = widgets.Text(description="Division:")

search_button = widgets.Button(description="Search")

# Create output widget
output_widget = widgets.Output()



def normalize_text(text):
    text = unicodedata.normalize("NFKD", text)  # Normalize to remove special characters
    text = re.sub(r'\s+', ' ', text)  # Replace multiple spaces with a single space
    text = text.strip()  # Remove leading/trailing spaces
    return text


df["Activity_Id"] = df["Activity_Id"].astype(str).apply(normalize_text)
df["ACT_IMPACTED"] = df["ACT_IMPACTED"].astype(str).apply(normalize_text)
df["ACT_DISCIPLINE"] = df["ACT_DISCIPLINE"].astype(str).apply(normalize_text)


# Define search function
def search_df(button):
  with output_widget:
    output_widget.clear_output()
    activity = str(activity_input.value)
    impacted_discipline=str(impacted_discipline_input.value)
    impacted_items = str(impacted_items_input.value)






    # Filter the DataFrame
    pattern_impacteditems = '|'.join(impacted_items.split(','))  # Create regex pattern for any of the impacted items
    pattern_impacteddiscipline = '|'.join(impacted_discipline.split(','))


    filtered_df = df[
        (df["Activity_Id"] == activity) &
        (df["ACT_IMPACTED"].str.contains(pattern_impacteditems, regex=True)) &
        (df["ACT_DISCIPLINE"].str.contains(pattern_impacteddiscipline, regex=True))
    ]



    if not filtered_df.empty:
          impacted_items_list = impacted_items.split(',')  # Split comma-separated values
          impacted_discipline_list = impacted_discipline.split(',')  # Split comma-separated values
          all_impacted_items = []
          all_impacted_discipline = []

          for index in filtered_df.index:
            all_impacted_items.extend(str(filtered_df.loc[index, 'ACT_IMPACTED']).split(','))  # Get all impacted items from filtered rows
            all_impacted_discipline.extend(str(filtered_df.loc[index, 'ACT_DISCIPLINE']).split(','))  # Get all impacted items from filtered rows

          # Find items not in impacted_items_list and not 'nan'
          remaining_missing_impact = [item for item in all_impacted_items if item not in impacted_items_list and item != 'nan']
          remaining_missing_impact = list(set(remaining_missing_impact))  # Remove duplicates

          remaining_missing_discipline = [item for item in all_impacted_discipline if item not in impacted_discipline_list and item != 'nan']
          remaining_missing_discipline = list(set(remaining_missing_discipline))  # Remove duplicates

          print(f"Remaining Missing Impacted Items are : {','.join(remaining_missing_impact)}")  # Join missing impacts with commas
          print(f"Remaining Missing Impacted Disciplines are : {','.join(remaining_missing_discipline)}")
    else:
          print("No matchings found.")



# search button

search_button.on_click(search_df)

In [113]:
display(activity_input, impacted_discipline_input, impacted_items_input, search_button, output_widget)

Text(value='', description='Activity Id:')

Text(value='', description='Disciplines impacted:')

Text(value='', description='Impacted Items:')

Button(description='Search', style=ButtonStyle())

Output()